src/io.rs

```diff
@@ -39,12 +40,11 @@ pub mod async_ {
     {
         let len_buf = {
             let mut len_buf = [0u8; std::mem::size_of::<u64>()];
-            match reader.read(&mut len_buf).await {
-                Ok(0) => return Ok(None),
-                Ok(n) if n == len_buf.len() => (),
-                Ok(_) => return Err(Error::UnexpectedEofError),
-                Err(error) => return Err(error.into()),
-            };
+            match reader.read_exact(&mut len_buf).await {
+                Err(ref e) if e.kind() == ErrorKind::UnexpectedEof => return Ok(None),
```

Comment:
It does not distinguish whether length is zero or not. They have different meanings here. If the length is zero, it hits an EOF with no more data, but if length is not zero, it ends with incomplete data. The function has to tell caller if there is more data available and whether it is complete.

Reply:
@jerry73204 It's quitely hard to distinguish these two situation. see this: [read_exact and EOF](https://internals.rust-lang.org/t/read-exact-and-eof/3047/2).
I think it's normal behavior just quit the processing when there is breaking. I also saw this from other implements like python: 
* [tfrecord-lite](https://github.com/jongwook/tfrecord_lite/blob/b5207374801445f763967ae75fb1f798b71be758/tfrecord_lite.pyx#L64)

Or we can check the filesize and the current position from tell (also breaks when using  stdin or other streaming method)